lgc/patch/PatchPeepholeOpt.cpp

```diff
@@ -132,9 +132,32 @@ void PatchPeepholeOpt::visitIntToPtr(IntToPtrInst &intToPtr) {
   if (!constOffset)
     return;
 
+  if (intToPtr.user_empty())
+    return;
+
+  auto *const user = cast<Instruction>(intToPtr.user_back());
+
+  const unsigned userOpcode = user->getOpcode();
+  // Check if user is Load or GEP instruction (right now only these two instructions are used).
+  if (userOpcode != Instruction::Load && userOpcode != Instruction::GetElementPtr)
+    return;
+
+  Type *elementType = nullptr;
+  if (auto loadInst = dyn_cast<LoadInst>(user)) {
+    elementType = loadInst->getType();
+
+  } else if (auto getElemPtr = dyn_cast<GetElementPtrInst>(user)) {
+    elementType = getElemPtr->getSourceElementType();
+
+  } else {
+    llvm_unreachable("Should never be called!");
+    return;
```

Comment:
You could have simplified this by removing the opcode check above.

Note that, under the hood, the `dyn_cast`s are really the exact same opcode comparisons, so doing the opcode check in advance has no performance benefit.

(This is why LLVM has the `dyn_cast` infrastructure which is used instead of C++'s native dynamic_cast.)

Reply:
I thought it will be more readable to have exit function check in line 142-143.

I will push new PR for the change that you are proposing (since this is merged)